accounts/abi/abi.go

```diff
@@ -101,52 +100,143 @@ func (abi ABI) Pack(name string, args ...interface{}) ([]byte, error) {
 }
 
 // toGoType parses the input and casts it to the proper type defined by the ABI
-// argument in t.
-func toGoType(t Argument, input []byte) interface{} {
+// argument in T.
+func toGoType(i int, t Argument, output []byte) (interface{}, error) {
+	index := i * 32
+
+	if index+32 > len(output) {
+		return nil, fmt.Errorf("abi: cannot marshal in to go type: length insufficient %d require %d", len(output), index+32)
+	}
+
+	// Parse the given index output and check whether we need to read
+	// a different offset and length based on the type (i.e. string, bytes)
+	var returnOutput []byte
+	switch t.Type.T {
+	case StringTy, BytesTy: // variable arrays are written at the end of the return bytes
+		// parse offset from which we should start reading
+		offset := int(common.BytesToBig(output[index : index+32]).Uint64())
+		if offset+32 > len(output) {
+			return nil, fmt.Errorf("abi: cannot marshal in to go type: length insufficient %d require %d", len(output), offset+32)
+		}
+		// parse the size up until we should be reading
+		size := int(common.BytesToBig(output[offset : offset+32]).Uint64())
+		if offset+32+size > len(output) {
+			return nil, fmt.Errorf("abi: cannot marshal in to go type: length insufficient %d require %d", len(output), offset+32+size)
+		}
+
+		// get the bytes for this return value
+		returnOutput = output[offset+32 : offset+32+size]
+	default:
+		returnOutput = output[index : index+32]
+	}
+
+	// cast bytes to abi return type
 	switch t.Type.T {
 	case IntTy:
-		return common.BytesToBig(input)
+		return common.BytesToBig(returnOutput), nil
 	case UintTy:
-		return common.BytesToBig(input)
+		return common.BytesToBig(returnOutput), nil
 	case BoolTy:
-		return common.BytesToBig(input).Uint64() > 0
+		return common.BytesToBig(returnOutput).Uint64() > 0, nil
 	case AddressTy:
-		return common.BytesToAddress(input)
+		return common.BytesToAddress(returnOutput), nil
 	case HashTy:
-		return common.BytesToHash(input)
+		return common.BytesToHash(returnOutput), nil
+	case BytesTy, FixedBytesTy:
+		return returnOutput, nil
+	case StringTy:
+		return string(returnOutput), nil
 	}
-	return nil
+	return nil, nil
```

Comment:
Instead of this nil/nil version, wouldn't adding a default branch to the switch be better. Also thinking that maybe we should rather return an error that `unknown type` instead of nil?